Collect sequences of ByteBufs into a single ByteBuf

jamesgorman2 opened the pull request:
Two methods are provided - collect all and collect up to max bytes. The
latter will emit an error if over max bytes are emitted from upstream.
#430

jamesgorman2 committed:
Collect sequences of ByteBufs into a single ByteBuf

Two methods are provided - collect all and collect up to max bytes.

NiteshKant commented on rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java:
```diff
@@ -0,0 +1,153 @@
+                            @Override
+                            public Observable<ByteBuf> call(ByteBuf buf) {
+                                return buf.readableBytes() > maxBytes ?
+                                        Observable.<ByteBuf>error(OnErrorThrowable.addValueAsLastCause(new TooMuchDataException(), buf)) :
```
`buf` should be released here as no one will access it after this.

jamesgorman2 replied:
see comment below on this

NiteshKant commented on rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java:
```diff
@@ -0,0 +1,153 @@
+
+        @Override
+        public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+            return upstream.scan(
```
If I understand this correctly, the `scan+skip+takeUntil` dance is required just to make sure we don't cross the buffer limit?
If so, then may be it is ok to just throw an `Exception` which will anyways be sent as an `onError` to the downstream, does that make sense? It will simplify the code.

NiteshKant replied:
@jamesgorman2 what's your take on this comment i.e. just do `collect()` instead of `scan+skip+takeUntil` and throw when the size limit is breached.

jamesgorman2 replied:
Sorry, notifications got buried.

I originally planned to just drop the data and throw, but I realised I sometimes want to know what I got before maxBytes was received (especially for logging - ie why am I getting too many bytes?).

This is making the error either a simple 'do not continue' or a signal that there was an overflow.

BUT, from your comment above it looks like there is a risk of frequent memory leak if users don't care about the content of the error or call `ContentSource#autoRelease()`.

On the basis of this, I think the only sensible path is roll back to a simpler fail-only model (ie what you are suggesting above).

NiteshKant replied:
Yes but you still need to `release` before throwing as nobody else will be able to access the last buffer and composite after this error.

jamesgorman2 replied:
This buf (the composite) was available from
`((OnErrorThrowable.OnNextValue) tooMuchDataException.getCause()).getValue()`
_if_ the user decides to actually get it. A diligent user could make sure they always check and release this, but it's a pretty massive hole since it's on the error path.

(there was also another buf that was not getting released that was the cause of the overflow - I think we are talking at cross purposes here through :-) )

I'm quickly putting together a proposal to close both these holes that follows your suggestion (and should close leaks!)

jamesgorman2 replied:
New version up should address these. The interface is not as nice since I can't get the bytes collected so far, but it is safer.

Also added a comment about (not) using autorelease since the whole thing dies if you do.

NiteshKant replied:
> > Also added a comment about (not) using autorelease since the whole thing dies if you do.

Hmm ... that is unfortunate 😞 but I don't think there is a way to avoid this easily.

jamesgorman2 replied:
I did some tests with the `AutoreleaseOperator` and it looked pretty much like the only way out would be to copy bytes so I left. 

It did give me some good thoughts on safe memory handling for an `Operator<ByteBuf, Char>`, so I'll look at that as a small win

NiteshKant commented on rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java:
```diff
@@ -0,0 +1,153 @@
+    /**
+     * Collect all emitted ByteBufs into a single ByteBuf.
+     */
+    public static Transformer<ByteBuf, ByteBuf> all() {
```
Thinking about this more, I think we can cap `all()` on `Integer.MAX_VALUE`. This will have sane limits (2GB) on the buffer and reuse the code for both methods, WDYT?

jamesgorman2 replied:
I'm neither here nor there, but this did make me realise there was an overflow bug calling `byteBuf.writerIndex(collector.byteBuf.writerIndex() + i)` if more that `Integer.MAX_VALUE` total bytes are received by `#upTo()` anyway so I rejigged the whole thing to be safer. 

It's a bit more complex inside but should be correct w.r.t. sizes.

jamesgorman2 committed:
Remove potential memory leak

We were returning the `ByteBuf` in the exception on failure, but this would
leak if not explicitly released by the user. There was also a memory leak in
failing to release the overflowing buf when more than maxBytes is read.

jamesgorman2 committed:
Update with warning about autorelease

NiteshKant commented on rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    private static class CountingCollector {
+        private long count;
```
I think `CompositeBuffer.readableBytes()` would suffice instead of keeping this count, which in turn removes the need of this class.

jamesgorman2 replied:
Good call, I'll simplify this.

NiteshKant commented on rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java:
```diff
@@ -0,0 +1,125 @@
+                        if (collector.count <= maxBytes) {
+                            int i = buf.readableBytes();
+                            collector.byteBuf.addComponent(buf);
+                            collector.byteBuf.writerIndex(collector.byteBuf.writerIndex() + i);
```
I think now `CompositeByteBuf` has an overload for `addComponent` which takes a `increaseWriterIndex` flag, which you can use.

jamesgorman2 replied:
Ditto for this one.

NiteshKant commented on rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java:
```diff
@@ -0,0 +1,125 @@
+                        } else {
+                            collector.byteBuf.release();
+                            buf.release();
+                            throw new TooMuchDataException();
```
Should we pass a message here specifying what was the limit and how much was received?

jamesgorman2 replied:
I wasn't sure about this - happy to put one in. Only makes sense to refer to maxBytes since we never end the stream.

jamesgorman2 committed:
Remove redundant code

jamesgorman2 committed:
add newline at end of file

jamesgorman2 commented:
I've done those updates (and improved the javadoc further). Are you running auto-squash merges or should I squash this down when you're happy with it?

as an aside, it's a pity that memory but management meant that some of the nice stuff I wanted couldn't work, but it is nice to remove code :-)

NiteshKant commented:
@jamesgorman2 yep I do squash and merge so you don't need to squash!

> > it's a pity that memory but management meant that some of the nice stuff I wanted couldn't work, but it is nice to remove code :-)

Tradeoffs!!
